### DISCUSSION TITLE
Is there a way to do fast forward / rewind with playback?

### DISCUSSION THREAD

**Discussion by 0x0fe** [pschatzmann/arduino-audio-tools#579] [Q&A] [answered]
everything is in the title, to be used with FF and rewind key when there is no UI, so the playback is the only feedback possible to know where we are at in the file.

**Comment by pschatzmann** [comment]
play back from where using which functionality ? 
There are about 100 different way to play audio with this framework....
What would FF exactly do ?

**Reply by 0x0fe** [reply]
for example when using the player from SDMMC

**Reply by pschatzmann** [reply]
If you look at the api you can position to any file using the index position...
You could play at any speed if this is what you mean with FF by using a ResampleStream

**Reply by 0x0fe** [reply]
i have seen set index, but this is not what i can use (because there is no UI). What i mean is accelerated playback, forward and backwrd, same as if you hold the FF or RW key on a CD player for example.

**Reply by pschatzmann** [reply]
I don't understand why you think you would you need a UI and there are also the next and prior methods for navigating back and forward.
I suggest that you look at [this example](https://github.com/pschatzmann/arduino-audio-tools/blob/main/examples/examples-audiokit/player-sdfat-audiokit/player-sdfat-audiokit.ino) which uses the buttons of the audiokit

**Reply by 0x0fe** [reply]
but, next and prior is to change to the next/previous file, right? what i mean is the standard transport function inside the file, to navigate the content of the file with an audio feedback

**Comment by pschatzmann** [comment, marked as answer]
Well you can use the regular file operatons to (re)set the actual file position....
I don't know what you mean with with an audio feedback

**Reply by 0x0fe** [reply]
mmh, no, you never used fast forward and fast rewind on an audio player? the playback is accelerated, and when using rewind it plays backward (also accelerate), files operation or setting playback position on the file are of no help to acheive that.

**Comment by pschatzmann** [comment]
No, I have never seen such a functionality!
Playing fast forward is feasible but playing audio backward does not make any sense !
What would make sense if moving the file pointer back and restart to play forward

**Comment by 0x0fe** [comment]
well, that is what most player devices do, when fast rewind with playback the audio is played in reverse direction, i believe that is also what CD players do. In fact i dont really see how you could do fast rewind with playback and play the audio in the forward direction.
Here is a sample from an MP3 player, when holding the FF and RW keys we can hear the fast playback, forward and backward. 
[transport_test.zip](https://github.com/pschatzmann/arduino-audio-tools/files/10467769/transport_test.zip)

**Reply by pschatzmann** [reply]
I am working with Linux and OSX and I am sorry, no this is not how things are working for me!
Anyhow you can achieve this by simple repositioning the actual file pointer and after you hit the end move to the next file

**Reply by 0x0fe** [reply]
well, it would not make any sense on a computer because you have an UI and the index is set with visual feedback on a timeline. But many simple audio devices with only 3 or 4 transport keys do not have any kind of UI and for the last 40 years this is how things have been done to navigate a long file / media, with audio feedback when using fast forward/rewind. Most MP3 SoC do implement this too. Moving the file pointer will maybe work in forward direction, but for reverse i have doubts.

**Reply by pschatzmann** [reply]
Yes, for everything which is file based this works. The good thing with MP3 is that if you do not reposition on a frame boundry the content is usually just ignored until you hit the next full frame. On WAV files you can move to any position which is a multiple of (bytes per sample * channels)
But it will not work on any medium which does not have any file repositioning fuctionality. (e.g. URLStream)

**Reply by 0x0fe** [reply]
yes, that is convenient, I am now trying to implement the FF and RW functions, so i will determine how many chunk should be read / skipped based on the file length and bitrate then perform file seeks at regular interval as long as the FF/RW keys are held, I wonder, where i should perform the seek operation? Is it in the copier? i currently use AudioSourceCallback rather than AudioSourceSDMMC due to the changes in indexing / file select methods.

**Reply by pschatzmann** [reply]
I suggest that you just perform the seek when pushing the button. The most generic way to access the current playing file is by calling the getStream() method on the player which returns a pointer to a Stream and if you use any file related streams you can just cast it to a File pointer.

**Reply by 0x0fe** [reply]
oh, i see, directly seeking audioFile from the keys handler, outside of the library, that makes sense since i manage the file directly now.

**Comment by 0x0fe** [comment]
ok, so the file seek works on the princple, now i need to find a way to performs very short fade ins/outs and also control how many chunks are copied/played between the seeks, so that the result sounds ok. At the moment it creates artifacts, presumably due to sharp amplitudes transitions outside zero crosses.

**Comment by pschatzmann** [comment]
We could expose the fade object in the Player with a public method

**Reply by 0x0fe** [reply]
yes that would be convenient, for now i experiment with player.setVolume, BTW i had to change the return type of player.copy to size_t to pass the count of copied bytes, so that i can know when chunks have been writen and increase my chunk counter only in this case.

Also, when using player.setVolume, since it is blocking and as far as i understand no operation can occur at the same time, if i change it in a loop, it will in effect only apply the last set volume for any given chunk, So the minimum "samples" number per volume step would be the buffer size of the copier, correct?

As a side note I feel on ESP32 the DMA block could be used for these copy operations, blocking delays() in player.copy are kinda problematic since this runs in the main loop (i will see that later).

**Reply by 0x0fe** [reply]
Oh, this is new, seems i should update to the lastest commit and use this, once i figure out how it works.

![image](https://user-images.githubusercontent.com/89607149/214813547-0f815b51-02f4-4366-92f7-73b978837891.png)

**Reply by 0x0fe** [reply]
![image](https://user-images.githubusercontent.com/89607149/214848312-c3ec4de1-0914-4072-b8ea-86f98f16805a.png)

I got this message when i start the stream, but i am not using fade here, is there something else i need to initilize with the new changes?

I have the notifiychange set

![image](https://user-images.githubusercontent.com/89607149/214856416-5a43eb3f-3628-4eb3-ba7a-a5cf3d2764d2.png)

So, it seems that this function should be called somewhere, maybe in AudioSourceCallback, but its not.

![image](https://user-images.githubusercontent.com/89607149/214858875-23144125-8720-4ffe-bd9d-fc73f2a19419.png)

**Reply by pschatzmann** [reply]
I am wondering what you do to write to the output before the mp3 could inform i2s and the player about the audio format. I suggest that you set the log level to debug.
Oh I see, so I suggest to remove your logic since this is automatically handled, and if you want to keep it you should not notify I2S but the player

**Reply by 0x0fe** [reply]
in fact i kep the same i did before and just changed the library, i only have 3 callabcks, init, select and next
![image](https://user-images.githubusercontent.com/89607149/214866348-7eac3b25-40f7-4bc1-931c-1763014e952b.png)

When i start playing, all i do is indexing files, selecting index 0 and player.begin

Note taht the select callback is assigned to source object, because there is no public access to this callback in player.

![image](https://user-images.githubusercontent.com/89607149/214866790-10dad844-128a-4c25-abe1-92d9f12d6763.png)

**Reply by 0x0fe** [reply]
>  if you want to keep it you should not notify I2S but the player

Ah, that makes sense, i change to player rather that I2S

Edit : it works now

**Comment by 0x0fe** [comment]
So from what i understand of the new player code, it should be able to just use player.setActive() to performe automatic fade-in and fade-outs when i need, so i implemented it that way, seekCnt counts the number of chunks copied by player.copy (increased when returns a value), so essentially i skip 5 chunk with file.seek, and play 150 chunks, i would like to perform the fade in and fade out on the first and last chunk of the played section (150 chunk). But it doesnt seems to work with player.setActive(). Should i proceed differently?

![image](https://user-images.githubusercontent.com/89607149/214868216-131982d8-671e-491b-b253-de6caa3be03b.png)

**Reply by pschatzmann** [reply]
I think the easiest would be to move the copy logic directly into your navigation:  call the copyN(), then fade out and then a final copy. You would need to have access to the copier though

**Reply by 0x0fe** [reply]
but the copy logic is ok, and i really hneed to avoid blocking the main loop, my question was more related to the fade in and fade out, is player.setActive appropriate here?  or should i call some lower level function from fade.h?

**Comment by pschatzmann** [comment]
I would really think that blocking the main loop might help.
In any case, try the easy thing first...

**Reply by 0x0fe** [reply]
i cannot block the main loop because several other unrelated functions are using it. I want to try performing the fade "manually" but i still havent figured how it should be done by looking at fade.h, i try to find some examples in the other files.

**Reply by mydroidbox** [reply]
Hi 0x0fe, I've been trying to acheive this ffwd and rewind for a couple days but
I confess the code is above my head. Could you share a copy of your final implementation.
That would greatly appreciated. Thanks.

**Reply by 0x0fe** [reply]
hello, sorry i missed the message, i implemented it yes, but its not working correctly still. I will tell here when it is working fine, as of now i have more major issues to solve with this because there is still farting noises when using the transport functions.

**Reply by bool-shark** [reply]
Thanks, appreciate it.

**Reply by 0x0fe** [reply]
so, i havent found a fully satisfactory way to implement this, but i'll give what i currently have and is the less worst so far, this implies the tact switches handler is in the same thread which serves player.copy, i dont show the key handler here because it is not useful, basically when key is just pressed it will call the transport function and when key is held if will call these seek functions instead (ff and rw).
Also something important : i had to modify the buffer sizes in the library wide preferences, because with default values it wil gllitch and produce very unpleasant noise artefacts, probably due to buffering issues.  As far as i can tell , APP and AUTO_CLEAR state have no influence on the seek function, but buffer size do.
The last player.copy is in a "for" statement because i tried to play a longer segment after the seek, however with the increased buffer size it ended up impacting the seek repeat time too much. So finally there is only one player.copy after seek, and the seeks is already quite sluggish.

```
#define DEFAULT_BUFFER_SIZE 2048
#define I2S_BUFFER_SIZE 1024
#define I2S_BUFFER_COUNT 20 
...
#define I2S_USE_APLL false 
#define I2S_AUTO_CLEAR false
```

```
void ff(void){  
  if(!digitalRead(KEY1)){
    player.setActive(0); // fade out
    player.copy();      
    audioFile.seek(audioFile.position()+(1024*30));      
    player.setActive(1); // fade in
    for(int i=0;i<1;i++)player.copy();  
  }  
}
void rw(void){
  if(!digitalRead(KEY3)){
    player.setActive(0); // fade out
    player.copy();      
    audioFile.seek(audioFile.position()-(1024*30));      
    player.setActive(1); // fade in
    for(int i=0;i<1;i++)player.copy();  
  }  
}
```
